R/coh_immunization.R

```diff
@@ -0,0 +1,344 @@
+  delta_limit <- immunization_delay
+
+  # Limit date
+  # If censoring date provided, use it. If not, use outcome date
+  limit_date <- data_set[[outcome_date_col]]
+
+  if (!is.null(censoring_date_col)) {
+    limit_date <- as.Date(ifelse(!is.na(data_set[[censoring_date_col]]),
+      yes = as.character(data_set[[censoring_date_col]]),
+      no = as.character(limit_date)
+    ))
+  }
+
+  # get difference with outcome date
+
+  data_set$imm_limit <- limit_date - delta_limit
+
+  # all other individuals' limit is set to end_cohort
+  data_set[is.na(data_set$imm_limit), "imm_limit"] <- end_cohort - delta_limit
+
+  # get differences from vaccination dates
+  cols_delta <- sprintf("delta_%i", seq_along(vacc_date_col))
+
+  data_set[cols_delta] <- lapply(vacc_date_col, function(col) {
+    vals <- as.numeric(data_set$imm_limit - data_set[[col]])
+    vals[vals < 0] <- NA
+    vals
+  })
+
+  # assign the lower value as delta_imm, keeping NAs where
+  # only NAs are present
+  # this `apply` replaces the earlier implementation and avoids
+  # a dependency
+  data_set$delta_imm <- apply(
+    # apply min over each row of a `data.frame` of the delta columns
+    # returning a vector of minimum delta values or NAs
+    # set data_set[, cols_delta] as dataframe to avoid errors when
+    # working with only one vaccination column
+    data.frame(data_set[, cols_delta]),
+    MARGIN = 1, FUN = function(x) {
+      #When several vaccines the function returns the first date
+      #that satisfies the limit date constraint if take_first = TRUE
+      #(i.e. the max delta). If take_firt = FALSE it returns the
+      #closest date to the event (i.e. the min delta).
+      if (all(is.na(x))) {
+        NA_real_
+      } else {
+        if (take_first) {
+          max(x, na.rm = TRUE)
+        } else {
+          min(x, na.rm = TRUE)
+        }
+      }
+    }
+  )
+
+  # immunization outcome as a vector
+  imm_out_date <- data_set$imm_limit - data_set$delta_imm +
+    immunization_delay
+
+  return(imm_out_date)
+}
+
+#' @title Return Column Name Associated to Immunizing Vaccine
+#'
+#' @description This function is used in cases where several columns with
+#' vaccine dates are provided by the user. It returns the name of the column
+#' of the vaccine used as immunizing. The parameters are set in agreement with
+#' `get_immunization_date`.
+#'
+#' @inheritParams make_immunization
+#' @return Name of the column taken as immunizing vaccine for each register.
+#' @keywords internal
+get_immunization_dose <- function(data_set,
+                                  immunization_date_col,
+                                  vacc_date_col,
+                                  immunization_delay) {
+  # The function can receive one or more column names in vacc_date_col
+  vacc_date_col_ <- vacc_date_col # hard coded to return right error message
+  for (vacc_date_col in vacc_date_col_) {
+    checkmate::assert_date(
+      data_set[[vacc_date_col]]
+    )
+  }
```

Comment:
When an error occurs, checkmate refers to `col` which does not have any meaning for the user, because they passed the parameter  `vacc_date_col` not `col`.